File: go/rewards/rewards_test.go
package rewards

import (
	"context"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"math/big"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync/atomic"
	"testing"
)

func realLogs(t *testing.T) []Log {
	t.Helper()
	b, err := os.ReadFile(filepath.Join("..", "..", "fixtures", "rewards_v4_logs.json"))
	if err != nil {
		t.Fatal(err)
	}
	var logs []Log
	if err := json.Unmarshal(b, &logs); err != nil {
		t.Fatal(err)
	}
	return logs
}

func w(n uint64) string { return fmt.Sprintf("%064x", n) }

func TestDecodeRealV4Log(t *testing.T) {
	l := realLogs(t)[0]
	e, ok := DecodeLog(l)
	if !ok || e.Version != 4 || len(e.Coin) != 42 {
		t.Fatalf("decoded %+v, ok=%v", e, ok)
	}
	data, _ := hex.DecodeString(strings.TrimPrefix(l.Data, "0x"))
	if want := "0x" + hex.EncodeToString(data[3*32-20:3*32]); e.Payouts[RoleCreator].Recipient != want {
		t.Errorf("creator = %s, want %s", e.Payouts[RoleCreator].Recipient, want)
	}
	if want := new(big.Int).SetBytes(data[7*32 : 8*32]); e.Payouts[RoleCreator].Currency.Cmp(want) != 0 {
		t.Errorf("creator currency = %v, want %v", e.Payouts[RoleCreator].Currency, want)
	}
	block, _ := strconv.ParseUint(strings.TrimPrefix(l.BlockNumber, "0x"), 16, 64)
	if e.Block != block {
		t.Errorf("block = %d, want %d", e.Block, block)
	}
}

func TestDecodeV3(t *testing.T) {
	payout, plat, trade := "0x"+strings.Repeat("11", 20), "0x"+strings.Repeat("22", 20), "0x"+strings.Repeat("33", 20)
	proto, usdc := strings.Repeat("44", 20), strings.Repeat("55", 20)
	l := Log{
		Address: "0x" + strings.Repeat("ab", 20), BlockNumber: "0x1c9c380", TransactionHash: "0x" + strings.Repeat("cd", 32), LogIndex: "0x5",
		Topics: []string{TopicTradeRewardsV3, AddressTopic(payout), AddressTopic(plat), AddressTopic(trade)},
		Data:   "0x" + strings.Repeat("0", 24) + proto + w(100) + w(25) + w(4) + w(20) + strings.Repeat("0", 24) + usdc,
	}
	e, ok := DecodeLog(l)
	if !ok || e.Version != 3 || e.Coin != "0x"+strings.Repeat("ab", 20) || e.Currency != "0x"+usdc {
		t.Fatalf("decoded %+v", e)
	}
	checks := map[Role][2]any{RoleCreator: {payout, int64(100)}, RolePlatformReferrer: {plat, int64(25)}, RoleTradeReferrer: {trade, int64(4)}, RoleProtocol: {"0x" + proto, int64(20)}}
	for role, want := range checks {
		p := e.Payouts[role]
		if p.Recipient != want[0] || p.Currency.Int64() != want[1] || p.Coin.Sign() != 0 {
			t.Errorf("%s = %+v, want %v", role, p, want)
		}
	}
}

func TestDecodeIgnoresOtherEvents(t *testing.T) {
	if _, ok := DecodeLog(Log{Topics: []string{"0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef"}, Data: "0x"}); ok {
		t.Fatal("decoded a Transfer as a reward")
	}
	if _, ok := DecodeLog(Log{}); ok {
		t.Fatal("decoded a log with no topics")
	}
}

func TestMissingRanges(t *testing.T) {
	got := missing(100, 200, [][2]uint64{{120, 130}, {125, 150}, {190, 250}})
	want := [][2]uint64{{100, 119}, {151, 189}}
	if fmt.Sprint(got) != fmt.Sprint(want) {
		t.Fatalf("missing = %v, want %v", got, want)
	}
	if got := missing(100, 200, [][2]uint64{{50, 300}}); len(got) != 0 {
		t.Fatalf("fully covered range reported gaps %v", got)
	}
}

// fakeNode serves eth_blockNumber and eth_getLogs from real logs, optionally refusing wide ranges.
func fakeNode(t *testing.T, logs []Log, head uint64, maxRange uint64, calls *atomic.Int32) *httptest.Server {
	t.Helper()
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		calls.Add(1)
		var req struct {
			Method string           `json:"method"`
			Params []map[string]any `json:"params"`
		}
		json.NewDecoder(r.Body).Decode(&req)
		switch req.Method {
		case "eth_blockNumber":
			fmt.Fprintf(w, `{"jsonrpc":"2.0","id":1,"result":"0x%x"}`, head)
		case "eth_getLogs":
			lo, _ := strconv.ParseUint(strings.TrimPrefix(req.Params[0]["fromBlock"].(string), "0x"), 16, 64)
			hi, _ := strconv.ParseUint(strings.TrimPrefix(req.Params[0]["toBlock"].(string), "0x"), 16, 64)
			if maxRange > 0 && hi-lo+1 > maxRange {
				fmt.Fprint(w, `{"jsonrpc":"2.0","id":1,"error":{"code":-32600,"message":"block range too large"}}`)
				return
			}
			var out []Log
			for _, l := range logs {
				b, _ := strconv.ParseUint(strings.TrimPrefix(l.BlockNumber, "0x"), 16, 64)
				if b >= lo && b <= hi {
					out = append(out, l)
				}
			}
			json.NewEncoder(w).Encode(map[string]any{"jsonrpc": "2.0", "id": 1, "result": out})
		}
	}))
	t.Cleanup(srv.Close)
	return srv
}

func TestScanFindsResumesAndShrinksChunks(t *testing.T) {
	logs := realLogs(t)
	first, _ := DecodeLog(logs[0])
	watch := first.Payouts[RoleCreator].Recipient
	var calls atomic.Int32
	node := fakeNode(t, logs, first.Block+500, 700, &calls) // refuses ranges over 700 blocks
	idx := NewIndexer(NewMemoryStore(), node.URL)
	from := first.Block - 1000
	n, err := idx.Scan(context.Background(), []string{"0x" + strings.ToUpper(watch[2:])}, ScanOptions{FromBlock: from})
	if err != nil {
		t.Fatal(err)
	}
	if n == 0 {
		t.Fatal("found no events for the creator of a real reward log")
	}
	events, _ := idx.Store.EventsFor([]string{watch}, 0)
	if len(events) != n {
		t.Fatalf("stored %d, reported %d", len(events), n)
	}
	before := calls.Load()
	again, err := idx.Scan(context.Background(), []string{watch}, ScanOptions{FromBlock: from})
	if err != nil || again != 0 {
		t.Fatalf("second scan found %d (err %v), want 0", again, err)
	}
	if calls.Load()-before > 1 { // only eth_blockNumber: every range is already scanned
		t.Fatalf("second scan made %d RPC calls, want 1", calls.Load()-before)
	}
}

func TestFileStoreRoundTrip(t *testing.T) {
	path := filepath.Join(t.TempDir(), "r.json")
	st, err := OpenFileStore(path)
	if err != nil {
		t.Fatal(err)
	}
	e, _ := DecodeLog(realLogs(t)[0])
	who := e.Payouts[RoleCreator].Recipient
	if err := st.Save([]*Event{e}, []string{who}, 4, 10, 20); err != nil {
		t.Fatal(err)
	}
	again, err := OpenFileStore(path)
	if err != nil {
		t.Fatal(err)
	}
	got, _ := again.EventsFor([]string{who}, 0)
	ranges, _ := again.Scanned(who, 4)
	if len(got) != 1 || got[0].Payouts[RoleCreator].Currency.Cmp(e.Payouts[RoleCreator].Currency) != 0 || fmt.Sprint(ranges) != "[[10 20]]" {
		t.Fatalf("round trip lost data: %v %v", got, ranges)
	}
}

func TestReportWithoutPricing(t *testing.T) {
	var events []*Event
	for _, l := range realLogs(t) {
		if e, ok := DecodeLog(l); ok {
			events = append(events, e)
		}
	}
	who := events[0].Payouts[RolePlatformReferrer].Recipient
	r, err := BuildReport(context.Background(), events, []string{who}, nil)
	if err != nil {
		t.Fatal(err)
	}
	if len(r.Lines) == 0 || r.Lines[0].Role != RolePlatformReferrer {
		t.Fatalf("lines = %+v", r.Lines)
	}
	if !strings.Contains(r.Text(), "Platform referral") || !strings.Contains(r.HTML("t"), "<svg") {
		t.Fatal("text or HTML rendering is missing its content")
	}
}

func TestFormatUSD(t *testing.T) {
	for in, want := range map[float64]string{0: "$0.00", 1234.5: "$1,234.50", 0.0042: "$0.0042", -2000: "$-2,000.00"} {
		if got := FormatUSD(in); got != want {
			t.Errorf("FormatUSD(%v) = %s, want %s", in, got, want)
		}
	}
}

File: typescript/test/rewards.test.ts
import { readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";

import { Indexer, MemoryStore, TOPIC_CREATOR_COIN_REWARDS, TOPIC_MARKET_REWARDS_V4, TOPIC_TRADE_REWARDS_V3, addressTopic, buildReport, decodeLog, formatUsd, missingRanges, type Log } from "../src/rewards/index.js";

const logs: Log[] = JSON.parse(readFileSync(new URL("../../fixtures/rewards_v4_logs.json", import.meta.url), "utf8"));
const w = (n: number) => n.toString(16).padStart(64, "0");
// Two real logs from one Base trade (tx 0x53f27c…f195): a CreatorCoinRewards payout for a creator coin and a
// CoinMarketRewardsV4 payout for a content coin, both to the same creator.
const cc: Log[] = JSON.parse(readFileSync(new URL("../../fixtures/creator_coin_rewards_logs.json", import.meta.url), "utf8"));
const CREATOR = "0xf4acf3edc65df843630976459ab1349a88258e6d";

describe("rewards", () => {
  it("decodes a real V4 log", () => {
    const e = decodeLog(logs[0]!)!;
    expect(e.version).toBe(4);
    expect(e.block).toBe(Number(BigInt(logs[0]!.blockNumber)));
    expect(e.payouts.creator.recipient).toBe("0x" + logs[0]!.data.slice(2 + 2 * 64 + 24, 2 + 3 * 64));
    expect(e.payouts.creator.currency).toBe(BigInt("0x" + logs[0]!.data.slice(2 + 7 * 64, 2 + 8 * 64)));
  });

  it("decodes V3 and ignores other events", () => {
    const [p, pl, t] = ["11", "22", "33"].map((b) => "0x" + b.repeat(20)) as [string, string, string];
    const e = decodeLog({ address: "0x" + "ab".repeat(20), blockNumber: "0x1c9c380", transactionHash: "0x" + "cd".repeat(32), logIndex: "0x5",
      topics: [TOPIC_TRADE_REWARDS_V3, addressTopic(p), addressTopic(pl), addressTopic(t)],
      data: "0x" + "0".repeat(24) + "44".repeat(20) + w(100) + w(25) + w(4) + w(20) + "0".repeat(24) + "55".repeat(20) })!;
    expect([e.version, e.payouts.platform_referrer.recipient, e.payouts.trade_referrer.currency]).toEqual([3, pl, 4n]);
    expect(decodeLog({ ...logs[0]!, topics: ["0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef"] })).toBeNull();
  });

  it("computes missing ranges", () => {
    expect(missingRanges(100, 200, [[120, 130], [125, 150], [190, 250]])).toEqual([[100, 119], [151, 189]]);
  });

  it("scans, resumes, and shrinks chunks a node refuses", async () => {
    const first = decodeLog(logs[0]!)!;
    const who = first.payouts.creator.recipient;
    let calls = 0;
    const node = (async (_: unknown, init: RequestInit) => {
      calls++;
      const req = JSON.parse(String(init.body));
      if (req.method === "eth_blockNumber") return Response.json({ jsonrpc: "2.0", id: 1, result: "0x" + (first.block + 500).toString(16) });
      const lo = Number(BigInt(req.params[0].fromBlock)), hi = Number(BigInt(req.params[0].toBlock));
      if (hi - lo + 1 > 700) return Response.json({ jsonrpc: "2.0", id: 1, error: { code: -32600, message: "block range too large" } });
      return Response.json({ jsonrpc: "2.0", id: 1, result: logs.filter((l) => { const b = Number(BigInt(l.blockNumber)); return b >= lo && b <= hi; }) });
    }) as typeof fetch;
    const store = new MemoryStore();
    const idx = new Indexer(store, "http://node", node);
    const n = await idx.scan([who.toUpperCase().replace("0X", "0x")], { fromBlock: first.block - 1000 });
    expect(n).toBeGreaterThan(0);
    expect(store.eventsFor([who])).toHaveLength(n);
    const before = calls;
    expect(await idx.scan([who], { fromBlock: first.block - 1000 })).toBe(0);
    expect(calls - before).toBe(1);
    const restored = MemoryStore.fromJSON(store.toJSON());
    expect(restored.eventsFor([who])[0]!.payouts.creator.currency).toBe(store.eventsFor([who])[0]!.payouts.creator.currency);
  });

  it("builds text and HTML reports without pricing", async () => {
    const events = logs.map(decodeLog).filter((e) => e !== null);
    const who = events[0]!.payouts.platform_referrer.recipient;
    const r = await buildReport(events, [who]);
    expect(r.lines[0]!.role).toBe("platform_referrer");
    expect(r.toText()).toContain("Platform referral");
    expect(r.toHtml()).toContain("<svg");
    expect([formatUsd(0), formatUsd(1234.5), formatUsd(0.0042)]).toEqual(["$0.00", "$1,234.50", "$0.0042"]);
  });

  it("decodes a real CreatorCoinRewards log (the creator's and protocol's shares on a creator-coin trade)", () => {
    expect(cc[0]!.topics[0]).toBe(TOPIC_CREATOR_COIN_REWARDS);
    const e = decodeLog(cc[0]!)!;
    expect([e.version, e.coin, e.currency]).toEqual([4, "0x3177fa60b8a342cd044badf34bf820c536094656", "0x1111111111166b7fe7bd91427724b487980afc69"]);
    expect(e.payouts.creator).toEqual({ recipient: CREATOR, currency: 11879451646867555805n, coin: 0n });
    expect(e.payouts.protocol.currency).toBe(11879451646867555805n);
    expect(e.payouts.platform_referrer.currency + e.payouts.trade_referrer.currency + e.payouts.doppler.currency).toBe(0n);
  });

  it("scans both V4 payout events, and a store saved before that rescans its V4 blocks once", async () => {
    const block = Number(BigInt(cc[0]!.blockNumber));
    const topicsAsked: unknown[] = [];
    let calls = 0;
    const node = (async (_: unknown, init: RequestInit) => {
      calls++;
      const req = JSON.parse(String(init.body));
      if (req.method === "eth_blockNumber") return Response.json({ jsonrpc: "2.0", id: 1, result: "0x" + (block + 10).toString(16) });
      topicsAsked.push(req.params[0].topics);
      const lo = Number(BigInt(req.params[0].fromBlock)), hi = Number(BigInt(req.params[0].toBlock));
      return Response.json({ jsonrpc: "2.0", id: 1, result: cc.filter((l) => { const b = Number(BigInt(l.blockNumber)); return b >= lo && b <= hi; }) });
    }) as typeof fetch;
    // what an earlier version saved: these blocks scanned for CoinMarketRewardsV4 only (no format field)
    const old = JSON.stringify({ events: [], scans: { [`${CREATOR}@v4`]: [[block - 5, block + 10]] } });
    const store = MemoryStore.fromJSON(old);
    const idx = new Indexer(store, "http://node", node);
    expect(await idx.scan([CREATOR], { fromBlock: block - 5 })).toBe(2);
    expect(topicsAsked.every((t) => JSON.stringify(t) === JSON.stringify([[TOPIC_MARKET_REWARDS_V4, TOPIC_CREATOR_COIN_REWARDS]]))).toBe(true);
    const again = MemoryStore.fromJSON(store.toJSON()); // saved now: its ranges count
    const before = calls;
    expect(await new Indexer(again, "http://node", node).scan([CREATOR], { fromBlock: block - 5 })).toBe(0);
    expect(calls - before).toBe(1); // only eth_blockNumber
  });
});
